api/v1beta1/machinedeployment_webhook.go

```diff
@@ -200,4 +200,11 @@ func PopulateDefaultsMachineDeployment(d *MachineDeployment) {
 	// Make sure selector and template to be in the same cluster.
 	d.Spec.Selector.MatchLabels[ClusterLabelName] = d.Spec.ClusterName
 	d.Spec.Template.Labels[ClusterLabelName] = d.Spec.ClusterName
+
+	if d.Spec.Template.Spec.InfrastructureRef.Namespace == "" {
+		d.Spec.Template.Spec.InfrastructureRef.Namespace = d.Namespace
```

Comment:
Can we make sure that after this change, a MachineDeployment isn't rolled out?

Reading the MachineDeployment controller code, seems that would be the case https://github.com/kubernetes-sigs/cluster-api/blob/095886906938c4a5491540c46a3e59e517a00a64/internal/controllers/machinedeployment/machinedeployment_sync.go#L145

Reply:
Maybe if it was deployed in the default namespace

Reply:
Okay, so if the namespace is not set we are unable to retrieve the referenced resources at all, correct? (it doesn't even fallback to default)

In this case I think it should be fine to ignore the pre-existing broken MachineDeployments that might exist.

Reply:
> Can we make sure that after this change, a MachineDeployment isn't rolled out?

Thanks for identifying the problem (MachineDeployment would get rolled out when the reference namespace is changed.).

However, if a user updates the reference, keeping the name, but changing the namespace, they would expect a rollout. (Maybe having reference in a different namespace is even supported, but that's at least _implied_ because the reference is an `ObjectReference`, not a `LocalObjectReference`).

> In alternative, we can do this kind of defaulting or enforcement only upon creation of new MachineDeployments for now

I like this idea.

Reply:
> In alternative, we can do this kind of defaulting or enforcement only upon creation of new MachineDeployments for now

Just that I get it right. So we implement and additional webhook which is only triggered on create? And defaulting the namespace won't be part of `PopulateDefaultsMachineDeployment`?

Can we maybe reconsider?

> Okay, so if the namespace is not set we are unable to retrieve the referenced resources at all, correct? (it doesn't even fallback to the default namespace)

So essentially my question is: are MachineDeployments without namespaces in refs working today? Because I think if they are not operational because the reconciliation just fails, we can keep the easier option to just add the defaulting.

Reply:
I liked the idea of defaulting this field only on create. But to do that, the webhook needs to know the event, which is in the AdmissionReview request, which isn't available to the `Default` method. We'd need to use a different API for the MachineDeployment webhook. (Another use case that needs the same approach is discussed in https://github.com/kubernetes-sigs/controller-runtime/issues/1600).

Reply:
> If I saw correctly the Namespace field of the refs is not used at all currently. The MachineDeployment, MachineSet and Machine controllers are all using the Namespace of the MachineDeployment, MachineSet and Machine instead.

Thanks for highlighting this @sbueringer!

So:
1. The namespace field is not read by the MachineDeployment controller.
2. The (undocumented?) requirement is that both referring and referenced objects must be in the same namespace.

To me, these two facts mean that we can safely apply the default on create and update. @vincepri WDYT?

Reply:
1.+2. Sounds about right. Even more. The MD, the MS and the Machine controller are all using the MD/MS/Machine namespace instead of the ones from the references.

Unfortunately Vince is still right :). When we propagate those newly set fields down to MS we calculate a hash over the MS, and when that hash changes we get a full rollout.

Reply:
Oh, right.

So what we probably want is to exclude the field from the hash. And we'd have to recalculate hashes in some way that doesn't cause a rollout. I think that's something  we would have to do whenever we replace `ObjectReference` with a type that doesn't have a namespace field.

Reply:
We could fix the problem only going forward for now, as in on creation, make sure that the namespace is forced and defaulted.

The hashing of MachineDeployment has probably be re-architected in a way that follow what KubeadmControlPlane has done, which is to do a more semantical diff, rather than a hash based one.

Reply:
Agree. I think something like just excluding the namespaces now isn't enough as we already have MachineSets with Namespaces out there with hashes calculated with the Namespace.

Reply:
I agree, too.

Given that the controllers don't use the namespace field, I don't think the effort of moving to a lower-level webhook API is worth doing. I can close the PR.